lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -0,0 +1,466 @@
+                            break;
+                        }
+                        let now = (self.performance.now() as f32).ms();
+                        let current_frame_time =
+                            now - time.animation_loop_start - time.since_animation_loop_started;
+                        if current_frame_time > FRAME_TIME_THRESHOLD {
+                            let msg1 =
+                                "Shaders compilation takes more than the available frame time.";
+                            let msg2 = "To be continued in the next frame.";
+                            debug!(self.logger, "{msg1} {msg2}");
+                            break;
+                        }
+                    }
+                    Err(err) => {
+                        if self.context.is_context_lost() {
+                            break;
+                        }
+                        let err_msg = err.blocking_report(&self.context);
+                        error!(self.logger, "{err_msg}");
+                    }
+                }
+            }
+        }
+    }
+
+    /// Runs the next compiler job if there is any left and if it will not cause too many jobs being
+    /// run in parallel. The result [`bool`] indicates if the call to this function did any
+    /// progress.
+    #[profile(Detail)]
+    fn run_step(&mut self) -> Result<bool, Error> {
+        let did_progress = |_| Ok(true);
+        let no_progress = Ok(false);
+        let jobs = &self.jobs;
+        if self.current_parallel_job_count() < MAX_PARALLEL_COMPILE_JOBS {
+            match () {
+                _ if !jobs.compile.is_empty() => did_progress(self.run_next_compile_job()?),
+                _ if !jobs.link.is_empty() => did_progress(self.run_next_link_job()?),
+                _ if !jobs.link_check.is_empty() => did_progress(self.run_next_link_check_job()?),
+                _ => {
+                    if jobs.khr_completion_check.is_empty() {
+                        debug!(self.logger, "All shaders compiled.");
+                        self.dirty = false;
+                    }
+                    no_progress
+                }
+            }
+        } else if !jobs.link_check.is_empty() {
+            did_progress(self.run_next_link_check_job()?)
+        } else {
+            let msg1 = "Maximum number of parallel shader compiler jobs.";
+            let msg2 = "Skipping spawning new ones.";
+            debug!(self.logger, "{msg1} {msg2}");
+            no_progress
+        }
+    }
+
+    /// Get the number of shader compilation jobs run in parallel. Please note, that it is
+    /// impossible to get separate values for compilation and linking jobs here, because checking
+    /// it is costly and can prevent the parallelism altogether. To learn more, see:
+    /// https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices#dont_check_shader_compile_status_unless_linking_fails
+    fn current_parallel_job_count(&self) -> usize {
```

Comment:
Doesn't include compiling state. `+ link.len()`?

Reply:
Hmmmm, that's right. I'm wondering how I missed that. Thanks